apollo-router/src/configuration/cors.rs

```diff
@@ -84,90 +154,338 @@ impl Cors {
+            .and_then(|oc| {
+                if oc.allow_headers.is_empty() {
+                    None
+                } else {
+                    Some(&oc.allow_headers)
+                }
+            })
+            .unwrap_or(&config.allow_headers);
+
+        let expose_headers = origin_config
+            .map(|oc| oc.expose_headers.as_ref())
+            .or(config.expose_headers.as_ref());
+
+        let methods = origin_config
+            .and_then(|oc| {
+                if oc.methods.is_empty() {
+                    None
+                } else {
+                    Some(&oc.methods)
+                }
+            })
+            .unwrap_or(&config.methods);
+
+        let max_age = config.max_age;
+
+        // Set Access-Control-Allow-Origin
+        if let Some(origin) = origin {
+            if config.allow_any_origin {
+                response.headers_mut().insert(
+                    ACCESS_CONTROL_ALLOW_ORIGIN,
+                    http::HeaderValue::from_static("*"),
+                );
+            } else if origin_config.is_some() {
+                // Only set the header if we found a matching origin configuration
+                response
+                    .headers_mut()
+                    .insert(ACCESS_CONTROL_ALLOW_ORIGIN, origin.clone());
+            }
+            // If no matching origin config found, don't set the header (origin will be rejected)
+        }
+
+        // Set Access-Control-Allow-Credentials
+        if allow_credentials {
+            response.headers_mut().insert(
+                ACCESS_CONTROL_ALLOW_CREDENTIALS,
+                http::HeaderValue::from_static("true"),
+            );
+        }
+
+        // Set Access-Control-Allow-Headers
+        if !allow_headers.is_empty() {
+            for header in allow_headers {
+                // Use the header value as-is to match test expectations
+                response.headers_mut().append(
+                    ACCESS_CONTROL_ALLOW_HEADERS,
+                    http::HeaderValue::from_str(header)
+                        .unwrap_or_else(|_| http::HeaderValue::from_static("")),
+                );
+            }
         } else {
-            Ok(cors.allow_origin(cors::AllowOrigin::list(parse_values(
-                &self.origins,
-                "origin",
-            )?)))
+            // If no headers are configured, mirror the client's Access-Control-Request-Headers
+            if let Some(request_headers) = request_headers {
+                if let Ok(headers_str) = request_headers.to_str() {
+                    response.headers_mut().insert(
+                        ACCESS_CONTROL_ALLOW_HEADERS,
+                        http::HeaderValue::from_str(headers_str)
+                            .unwrap_or_else(|_| http::HeaderValue::from_static("")),
+                    );
+                }
+            }
+        }
+
+        // Set Access-Control-Expose-Headers
+        if let Some(headers) = expose_headers {
```

Comment:
This header is only for NON preflight requests

https://github.com/tower-rs/tower-http/blob/35a6cb83242b2004352a8a45f97c0c2cb5475254/tower-http/src/cors/mod.rs#L693